overlord/hookstate/context.go

```diff
@@ -127,3 +128,50 @@ func (c *Context) Get(key string, value interface{}) error {
 
 	return nil
 }
+
+// State returns the state contained within the context
+func (c *Context) State() *state.State {
+	return c.task.State()
+}
+
+// Cached returns the cached value associated with the provided key.
+// It returns nil if there is no entry for key.
+func (c *Context) Cached(key interface{}) interface{} {
+	contextCache := c.task.State().Cached("hook-context")
+	cache, ok := contextCache.(map[interface{}]interface{})
+	if !ok {
+		return nil
+	}
+
+	return cache[key]
+}
+
+// Cache associates value with key. The cached value is not persisted.
+func (c *Context) Cache(key, value interface{}) {
+	contextCache := c.task.State().Cached("hook-context")
+	cache, ok := contextCache.(map[interface{}]interface{})
+	if !ok {
+		cache = make(map[interface{}]interface{})
+	}
+	cache[key] = value
+	c.task.State().Cache("hook-context", cache)
+}
+
+// OnDone requests the provided function to be run once the context knows it's
+// complete. This can be called multiple times; each function will be called in
+// the order in which they were added.
+func (c *Context) OnDone(f func() error) {
+	c.onDone = append(c.onDone, f)
```

Comment:
are we piggy backing on the state lock here? I'm a bit worried that this will not check that it's held though